storage/client_raft_test.go

```diff
@@ -541,6 +541,119 @@ func TestStoreRangeReplicate(t *testing.T) {
 	}
 }
 
+// TestStoreRangeDownReplicate verifies that the replication queue will notice
+// over-replicated ranges and remove replicas from them.
+func TestStoreRangeDownReplicate(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	mtc := startMultiTestContext(t, 5)
+	defer mtc.Stop()
+
+	store1 := mtc.stores[0]
+
+	// Split off a range from the initial range for testing; there are
+	// complications if the metadata ranges are removed from store 1, this
+	// simplifies the test.
+	splitKey := proto.Key("m")
+	rightKey := proto.Key("z")
+	{
+		replica := store1.LookupReplica(proto.KeyMin, nil)
+		mtc.replicateRange(replica.Desc().RangeID, 0, 1, 2)
+		desc := replica.Desc()
+		splitArgs := adminSplitArgs(splitKey, splitKey, desc.RangeID, store1.StoreID())
+		if _, err := replica.AdminSplit(splitArgs, desc); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Replicate the split range to all five stores.
+	replica := store1.LookupReplica(rightKey, nil)
+	desc := replica.Desc()
+	mtc.replicateRange(desc.RangeID, 0, 3, 4)
+
+	// Initialize the gossip network.
+	var wg sync.WaitGroup
+	wg.Add(len(mtc.stores))
+	key := gossip.MakePrefixPattern(gossip.KeyStorePrefix)
+	mtc.stores[0].Gossip().RegisterCallback(key, func(_ string, _ []byte) { wg.Done() })
+	for _, s := range mtc.stores {
+		s.GossipStore()
+	}
+	wg.Wait()
+
+	// Function to see if the replication level is appropriate.
+	type storeIDset map[proto.StoreID]struct{}
+	checkReplication := func() (bool, storeIDset) {
+		// Query each store for a replica of the range, generating a real map of
+		// the replicas.
+		foundReplicas := make([]*storage.Replica, 0, len(mtc.stores))
+		foundIDset := make(storeIDset)
+		for _, s := range mtc.stores {
+			r := s.LookupReplica(splitKey, nil)
+			if r != nil {
+				foundReplicas = append(foundReplicas, r)
+				foundIDset[s.Ident.StoreID] = struct{}{}
+			}
+		}
+
+		// Fail immediately if there are less than three replicas.
+		replicaCount := len(foundReplicas)
+		if replicaCount < 3 {
+			t.Fatalf("Removed too many replicas; expected at least three replicas, found %d", replicaCount)
+		}
+
+		// Ensure the local range descriptors everywhere agree with reality.
+		for _, repl := range foundReplicas {
+			replIDset := make(storeIDset)
+			for _, r := range repl.Desc().Replicas {
+				replIDset[r.StoreID] = struct{}{}
+			}
+			if !reflect.DeepEqual(replIDset, foundIDset) {
+				return false, replIDset
+			}
+		}
+
+		// If we have only three replicas, exit the loop.
+		if replicaCount == 3 {
+			return true, nil
+		}
+		return false, foundIDset
+	}
+
+	maxTimeout := time.After(5 * time.Second)
```

Comment:
I was and it really doesn't work well here.  This needs to do work to occasionally push it forward (forcing queues to run), and the exponential backoff means that it thrashes early on and is too slow later on.  I need the retries to be evenly spaced.